### PR TITLE
[Core] extend mpc python interface

### PR DESCRIPTION
**Pull request by armingeiser**
@KratosMultiphysics/interface-committee 

With these functions it is possible to get the information of existing mpcs (master and slave dofs, and corresponding weights) at python level.

### CODE REVIEW

File: kratos/python/add_constraint_to_python.cpp [hunk 1]
```diff
@@ -114,6 +114,10 @@ void AddConstraintToPython(pybind11::module &m)
     .def("SetValue", SetValueHelperFunctionMasterSlaveConstraint< Variable< std::string > >)
     .def("GetValue", GetValueHelperFunctionMasterSlaveConstraint< Variable< std::string > >)
 
+    .def("GetMasterDofsVector", &MasterSlaveConstraint::GetMasterDofsVector, pybind11::return_value_policy::reference_internal)
```
Comment (philbucher) [inline kratos/python/add_constraint_to_python.cpp:117]: Why is the special pybind return value policy necessary?

Comment (armingeiser) [reply kratos/python/add_constraint_to_python.cpp:117]: Otherwise the dofs get deleted when the python vector gets out of scope (at least it segfaults...)

Comment (armingeiser) [reply kratos/python/add_constraint_to_python.cpp:117]: As far as i see pybind creates a python list with the pointers inside.
```
[<Kratos.Dof object at 0x7ffa55a9cb30>, <Kratos.Dof object at 0x7ffa55a9cb70>, ...]
```

I think it is the same reason why e.g. it is needed for the NodesArray() in the model part has this policy:
```cpp
.def("NodesArray", &ModelPart::NodesArray, py::return_value_policy::reference_internal)
.def_property("Nodes", ModelPartGetNodes1, ModelPartSetNodes1)
.def("GetNodes", ModelPartGetNodes1)
```
where `NodesArray` returns the actual container, while `Nodes` gives the PointerVectorSet. but to be honest i am lost in the typdefs.

Comment (armingeiser) [reply kratos/python/add_constraint_to_python.cpp:117]: EDIT: according to pybind docs this is also the default for `def_property` getters.

**Comment by philbucher** [comment]
Ok fine for me 👍

**Comment by adityaghantasala** [comment]
Dofs are very deep in the code, i am not sure if exposing dofs to python is good. Especially now that dofs are raw pointers. 

Also does pybind still expose aS a shared pointer ?

**Comment by armingeiser** [comment]
After discussion with @adityaghantasala:
@pooyan-dadvand @RiccardoRossi @KratosMultiphysics/technical-committee Is there a technical  reason the dof object should not be exposed to python? E.g. in https://github.com/KratosMultiphysics/Kratos/blob/c937adcd0bf8344d96a3cd6975f36c9d16b241a3/kratos/python/add_mesh_to_python.cpp#L498
it seems there was an attempt, but is commented out.

**Comment by pooyan-dadvand** [comment]
If I remember correctly, @RiccardoRossi may correct me, we commented that one from the time that we tied the Dof data to the Nodal Data. This dependency made the creation of Dof without a node impossible. So we took it out from the Python interface to avoid having inconsistence dofs.

If we want to turn it back there, I would close anyway that one can create a dof by not exposing the constructor and even the copy constructor. Also the memory management should be held by cpp and not python.

**Comment by armingeiser** [comment]
> If we want to turn it back there, I would close anyway that one can create a dof by not exposing the constructor and even the copy constructor. Also the memory management should be held by cpp and not python.

Thats what is done in this PR. The Dof can neither be constructed not copied in python. And the vector of dofs that is exposed, has the `reference_internal` policy that keeps the ownership in cpp. Basically it is a read-only interface to the Dof.

**Comment by adityaghantasala** [comment]
@armingeiser  that clears my objections !

**Comment by armingeiser** [comment]
So the question remains where the dof object should be exposed. `add_node_to_python.cpp` or `add_mesh_to_python.cpp`.  We could even make a dedicated file `add_dof_to_python.cpp` file, so it can be found easily.

**Comment by armingeiser** [comment]
In #6099 the dofs python interface has been added. This MR is now only about extending the mpcs. The above discussions about the dof object are therefor not relevant for this MR anymore. IT would be nice if someone could approve.

**Review by philbucher — APPROVED** [review]

**Comment by adityaghantasala** [comment]
The node's GetDof is also a part of this PR. Iun related. But its ok.

**Comment by armingeiser** [comment]
> The node's GetDof is also a part of this PR. Iun related. But its ok.

Thats only a whitespace change from the merge commit.

**Pull request merged**